src/ddsrt/src/events/darwin/events_darwin.c

```diff
@@ -0,0 +1,236 @@
+  assert(kevent(queue->kq, &kev, 1, NULL, 0, NULL) != -1);
+#endif /* !LWIP_SOCKET */
+
+  /*create kevents array*/
+  queue->kevents = ddsrt_malloc(sizeof(struct kevent) * queue->cevents);
+  assert(queue->kevents);
+
+  return DDS_RETCODE_OK;
+}
+
+/**
+* @brief Finishes an event queue.
+*
+* Will free created containers and do any additional cleanup of things created in ddsrt_event_queue_init.
+*
+* @param[out] queue The queue to finish.
+*
+* @returns DDS_RETCODE_OK if everything went OK.
+*/
+static dds_return_t ddsrt_event_queue_fini(ddsrt_event_queue_t* queue) {
+  assert(queue);
+  /*deregister all currently stored triggers from kqueue*/
+  struct kevent kev;
+  for (unsigned int i = 0; i < queue->nevents; i++) {
+    ddsrt_event_t* evt = queue->events[i];
+    if (evt->flags & DDSRT_EVENT_FLAG_READ)
+      EV_SET(&kev, evt->data.socket.sock, EVFILT_READ, EV_DELETE, 0, 0, evt);
+    assert(kevent(queue->kq, &kev, 1, NULL, 0, NULL) != -1);
+  }
+#if !defined(LWIP_SOCKET)
+  EV_SET(&kev, queue->interrupt[0], EVFILT_READ, EV_DELETE, 0, 0, 0);
+  assert(kevent(queue->kq, &kev, 1, NULL, 0, NULL) != -1);
+  close(queue->interrupt[0]);
+  close(queue->interrupt[1]);
+#endif /* !LWIP_SOCKET */
+  close(queue->kq);
+
+  ddsrt_mutex_destroy(&queue->lock);
+  ddsrt_free(queue->events);
+  ddsrt_free(queue->kevents);
+  return DDS_RETCODE_OK;
+}
+
+ddsrt_event_queue_t* ddsrt_event_queue_create(void) {
+  ddsrt_event_queue_t* returnptr = ddsrt_malloc(sizeof(ddsrt_event_queue_t));
+  assert(returnptr);
+  ddsrt_event_queue_init(returnptr);
+  return returnptr;
+}
+
+dds_return_t ddsrt_event_queue_destroy(ddsrt_event_queue_t* queue) {
+  assert(queue);
+  ddsrt_event_queue_fini(queue);
+  ddsrt_free(queue);
+  return DDS_RETCODE_OK;
+}
+
+size_t ddsrt_event_queue_nevents(ddsrt_event_queue_t* queue) {
+  size_t ret;
+  ddsrt_mutex_lock(&queue->lock);
+  ret = queue->nevents;
+  ddsrt_mutex_unlock(&queue->lock);
+  return ret;
+}
+
+dds_return_t ddsrt_event_queue_wait(ddsrt_event_queue_t* queue, dds_duration_t reltime) {
+  assert(queue);
+
+  /*reset triggered status*/
+  ddsrt_mutex_lock(&queue->lock);
+  queue->ievents = 0;
+  for (unsigned int i = 0; i < queue->nevents; queue++)
+    ddsrt_atomic_st32(&queue->events[i]->triggered, DDSRT_EVENT_FLAG_UNSET);
+  ddsrt_mutex_unlock(&queue->lock);
+
+  struct timespec tmout = { reltime / 1000000000,     /* waittime (seconds) */
+                            reltime % 1000000000 }; /* waittime (nanoseconds) */
+  int nevs = kevent(queue->kq, NULL, 0, queue->kevents, (int)queue->nevents, &tmout);
+  ddsrt_mutex_lock(&queue->lock);
+  for (int i = 0; i < nevs; i++) {
+    ddsrt_event_t* evt = queue->kevents[i].udata;
+    if (NULL != evt) {
+      ddsrt_atomic_st32(&evt->triggered, DDSRT_EVENT_FLAG_READ);
+    }
+    else {
+#if !defined(LWIP_SOCKET)
+      char buf;
+      if (1 != read(queue->interrupt[0], &buf, 0))
+        return DDS_RETCODE_ERROR;
```

Comment:
You mustn't return without unlocking the queue first. (I apologise on C's behalf for it being obsolete since the late '50s.)

Reply:
Already saw this as being problematic after running the unittest on a mac.
Will be fixed.

Reply:
This is now implemented.